File: cpp/094.Binary Tree Inorder Traversal.cpp
/**
 * Source: LeetCode  094. Binary Tree Inorder Traversal
 * Author: ceezyyy
 * Date: 10-04-2019
 */


/*
Medium

Given a binary tree, return the inorder traversal of its nodes' values.

Example:

Input: [1,null,2,3]
   1
	\
	 2
	/
   3

Output: [1,3,2]

Follow up: Recursive solution is trivial, could you do it iteratively?
	
*/


#include<vector>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
	struct TreeNode {
		int val;
		TreeNode *left;
		TreeNode *right;
		TreeNode(int x) : val(x), left(NULL), right(NULL) {}
	};
    vector<int> res;
    vector<int> inorderTraversal(TreeNode* root) {
        inorderTraversal_Recursion(root);
        return res;
    }
    
    void inorderTraversal_Recursion(TreeNode *node) {
        if(node)
        {
            if(node->left) {
                inorderTraversal_Recursion(node->left);
            }
            res.push_back(node->val);
            if(node->right) {
                inorderTraversal_Recursion(node->right);
            }
        }
    }
};


/**
 * Runtime: 4 ms, faster than 59.48% of C++ online submissions for Binary Tree Inorder Traversal.
 * Memory Usage: 9.5 MB, less than 33.00% of C++ online submissions for Binary Tree Inorder Traversal.
 */

File: Python3/094.Binary Tree Inorder Traversal.py
"""
@Title: 094. Binary Tree Inorder Traversal
@Tag: binary tree
@Date: Mar-03 2020
@Author: ceezyyy
@Difficulty: Medium
"""


# Approach 1: (Iterative)


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        res, stack = [], [(root, False)]
        while stack:
            node, visited = stack.pop()  # the last element
            if node:
                if visited:
                    res.append(node.val)
                else:  # inorder: left -> root -> right
                    stack.append((node.right, False))
                    stack.append((node, True))
                    stack.append((node.left, False))
        return res


"""
Runtime: 24 ms, faster than 87.53% of Python3 online submissions for Binary Tree Inorder Traversal.
Memory Usage: 12.8 MB, less than 100.00% of Python3 online submissions for Binary Tree Inorder Traversal.
"""


"""
Time Complexity: O(n)
Space Complexity: O(n)
"""


# Approach 2: (Recursive)


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None


class Solution:
    def inorderTraversal(self, root: TreeNode) -> List[int]:
        res = []
        self.dfs(root, res)
        return res

    def dfs(self, node, res):
        if node:
            self.dfs(node.left, res)
            res.append(node.val)
            self.dfs(node.right, res)


"""
Runtime: 20 ms, faster than 97.04% of Python3 online submissions for Binary Tree Inorder Traversal.
Memory Usage: 12.8 MB, less than 100.00% of Python3 online submissions for Binary Tree Inorder Traversal.
"""


"""
Time Complexity: O(n)
Space Complexity: O(n)
"""
